boutiques_descriptors/micapipe_0_2_2.json

```diff
@@ -0,0 +1,713 @@
+{
+  "name":           "micapipe",
+  "tool-version":   "0.2.2",
+  "container-image": {
+    "image": "micalab/micapipe:v0.2.2",
+    "index": "docker://",
+    "type": "singularity"
+  },
+  "schema-version": "0.5",
+  "author":         "Multimodal Imaging and Connectome Analysis (MICA) Lab ",
+  "url":            "https://github.com/MICA-MNI/micapipe",
+  "description":    "Please cite: Cruces RR, Royer J, Herholz P, Lariviere S, Vos de Wael R, Paquola C, Benkarim O, Park BY, Degre-Pelletier J, Nelson MC, DeKraker J, Leppert IR, Tardif C, Poline JB, Concha L, Bernhardt BC. Micapipe: A pipeline for multimodal neuroimaging and connectome analysis. Neuroimage. 2022 Nov;263:119612. doi: 10.1016/j.neuroimage.2022.119612. Epub 2022 Sep 5. PMID: 36070839. doi: 10.1016/j.neuroimage.2022.119612",
+  "command-line": "micapipe [BIDS_DATASET] [OUTPUT_DIR] [DERIVATIVES] [SUB] [SES] [FS_LICENCE] [PROC_STRUCTURAL] [PROC_SURF] [POST_STRUCTURAL] [PROC_FLAIR] [PROC_FUNC] [PROC_DWI] [SC] [MPC] [GD] [QC] [QC_SUBJ] [CLEANUP] [ATLAS] [FREESURFER] [T1WSTR] [UNI] [MF] [SURF_DIR] [T1] [REGSYNTH] [TRACTS] [KEEP_TCK] [AUTOTRACT] [THREADS] [TMPDIR] [ACQSTR] [DWI_MAIN] [DWI_RPE] [DWI_PROC] [RPE_ALL] [REGAFFINE] [DWIACQ] [BTHR] [NO_BVAL] [WSC] [TCK] [FLAIRSTR] [MPC_ACQ] [MPC_IMG] [MPC_REG] [FUNC_MAIN] [FUNC_PE] [FUNC_RPE] [TOPUP] [SESANAT] [FIX] [NOFIX] [NSR] [GSR] [DROPTR] [NOFC]",
+  "inputs": [
+      {
+        "description": "Path to the BIDS Dataset directory",
+        "id": "subject_dir",
+        "name": "BIDS directory",
+        "optional": false,
+        "type": "File",
+        "value-key": "[BIDS_DATASET]",
+        "command-line-flag": "-bids"
+      },
+      {
+        "description": "Path to the derivatives folder",
```

Comment:
There is some sort of confusion here between [OUTPUT_DIR] and [DERIVATIVES]; the descriptor seems to say that DERIVATIVES is a necessary input, something that must be given to the tool before it starts. And OUTPUT_DIR seems to indicate that the tool will create, in the current working directory, a new folder called '[SUBJECT_DIR]_res', completely independently of whatever DERIVATIVES say.

I suspect DERIVATIVES is supposed to be a STRING which will create a folder, and then OUTPUT_DIR's template is supposed to be `[DERIVATIVES]/[SUBJECT_DIR]_res` maybe ?

Reply:
you are right. ` OUTPUT_DIR` is `DERIVATIVES`, I copied the template and I didn't change that. the flag `-out` should be the path to the output directory where the pipeline with create another directory to store all the subjects. e.g. <output>/micapipe_v0.2.0/[SUBJECT_DIR]. I don't really know how to fix it and make it compatible with CBRAIN

Reply:
Question. The directory provided with -out, does the tool create it automatically if it doesn't exist? Or does it have to exist already?

1. if it is created automatically, I suggest you a) change the input type to 'String' instead of 'File' at line 29 and b) change the path-template for output specification (line 559 and 560). The output specification doesn't have to have a value-key at all within the command, it is a specification of where to find the output, even if the command line itself doesn't have an explicit name on it.
2. if the directory si not created automatically, then change the command-line at line so that is starts with something like `mkdir [DERIVATIVES]; micapipe [BIDS_DATASET] ...`

Reply:
the directory provided must exist, the other one will be created by the pipeline to store all subjects `<out>/micapipe_v0.2.0/[SUBJECT_DIR]`. The outputs are in BIDS as well so it will follow the same structure.

Reply:
There' still something I don't get. The command line template is

```
micapipe [BIDS_DATASET] [OUTPUT_DIR] [DERIVATIVES]
```

based on the rest of the descriptor, the actual command invoked can be something like

```
micapipe mybidsdataset out -out deriv -sub sub-123
```

Assuming 'deriv' is a folder that exist, what does the tool create exactly? Is it `deriv/micapipe_v0.2.0/sub-123` ? Then what does the string "out" do? (provided by the value-key specified at line 559) What is `[OUTPUT_DIR]` used for?

Reply:
That is something I don't know how to do on the boutique file DERIVATIVES is the same as OUTPUT_DIR but instead of being a positional argument it requires the flag -out but it is the same. 
The pipeline will create the directory 'micapipe_v.0.2.0' inside the output directory that must exist. And it will store each subject's derivatives there like micapipe_v0.2.0/sub-###.

Reply:
Use bosh to see what commands you generate. E.g. create a small separate json file with values for your input, like this own (which I just created and verified that it works):

Simple JSON input file `input.json` for descriptor
```json
{
  "subject_dir": "subdirvalue",
  "derivatives": "dervalue",
  "fs_licence": "fake_license"
}
```

Then run the command `bosh exec simulate -i micapipe_0_2_2.json input.json`. In the current situation, I get this command (which is obviously wrong):

```
> bosh exec simulate micapipe_0_2_2.json -i i.json 
Generated Command:
micapipe -bids subdirvalue '[SUBJECT_DIR]_res' -out dervalue -fs_licence fake_license -atlas schaefer-400 -T1wStr T1w.nii -mf 3 -tracts 10M -threads 10 -tmpDir /tmp -b0thr 61 -flairScanStr FLAIR
```

Bosh will pick/insert random strings and numbers for the other optional inputs.

Remember that the templating system for the command is straighforward. If you want to join two values for ABC and DEF on the command-line, you can write `[ABC]/[DEF]` bosh will leave the slash exactly as-is.

Reply:
Updated the `bosh exec simulate` command above, I had put the `-i` at the wrong place.

Reply:
You removed too much. The special CBRAIN extension modules were good and necessary. But more than that, you also removed completely the output-files section of the descriptor, which means you are basically saying your tool doesn't produce any outputs at all. Remember, the point of this descriptor is to provide a formal description of how your tool works with files, so that an automated system (CBRAIN) can run it on behalf of users, in environments completely outside the users' control. So CBRAIN needs to know where the output files are located once the tool is finished, so it can transfer them back to the user. A `output-files` entry should contain the path of what to save, presumably using any of the input strings to create that path. From what I see n the descriptor, the "path-template" shoudl be something like `[OUTPUT_DIR]/micapipe_v0.2.0/[SUB]` (because OUTPUT_DIR is what the user selects for the location of the outputs, and you said the program creates 'micapipe_v0.2.0' under it, and then anotehr folder called [SUB] under that)